Label the projects created by cluster-loader

hongkailiu opened the pull request:
This feature would be useful for manual clean-up, such as `oc delete project -l test=purpose` if we have the labels defined in the config file:

```
provider: local
ClusterLoader:
  cleanup: false
  projects:
    - num: 2
      basename: clusterproject
      labels:
         purpose: test
```

hongkailiu commented:
/assign @mfojtik

hongkailiu commented:
/retest

hongkailiu commented:
/test extended_conformance_install

hongkailiu commented:
/retest

hongkailiu commented:
/retest

hongkailiu commented:
/test verify

mffiedler commented:
@sjug PTAL

sjug commented on test/extended/cluster/cl.go:
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
 	}
 }
 
+func labelNamespace(c kclientset.Interface, ns string, pairs map[string]string) (*v1.Namespace, error) {
```
Please move this into `utils.go`, you can see there are a host of dependencies being imported here (in this commit) that for the most part already are in utils.

The function name is unclear to me... Are we trying to `SetNamespaceLabels` or `UpdateNamespaceLabels`?

hongkailiu replied:
Done.
Moved and renamed to SetNamespaceLabels

sjug commented on test/extended/cluster/cl.go:
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
+	if len(pairs) == 0 {
+		return nil, nil
+	}
+	var d struct{ Metadata metav1.ObjectMeta `json:"metadata"`}
```
I've never seen a nested struct created like this, but I think the underlying issue is that you're trying to take existing golang types and then convert them to json which is unnecessary. The `Namespace.Labels` are already of type `map[string]string` which is the same type you've added to the `ClusterLoaderType`.

sjug commented on test/extended/cluster/cl.go:
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
+	}
+	var d struct{ Metadata metav1.ObjectMeta `json:"metadata"`}
+	d.Metadata = metav1.ObjectMeta{Labels:pairs}
+	bytes,err := json.Marshal(d)
```
This is totally unecessary.

sjug commented on test/extended/cluster/cl.go:
```diff
@@ -202,6 +208,19 @@ func newProject(nsName string) *projectapi.Project {
+	if err != nil {
+		return nil, err
+	}
+	return c.CoreV1().Namespaces().Patch(ns, types.MergePatchType, bytes)
```
An easy way to do this with native objects would be to get the existing namespace pointer with:
```
CoreV1().Namespaces().Get(ns, metav1.GetOptions{})
```
Modify the `Namespace.Label` directly by appending or setting the value to `pairs`.

Then just update the namespace:
```
CoreV1().Namespaces().Update(Namespace)
```
where `Namespace` is the modified object you recieved from the first `Get` and which you updated with `pairs`.

hongkailiu replied:
Done.
With a little concern/question:
Update method uses the whole ns object as the posting body while patch method uses only the label/metadata part. Probably the latter is a bit more efficient.

I tried "oc label project ... --loglelvel=10" and saw only metadata is in the body of http-req.
That was why I used patch method in the first commit here.

Maybe namespaces/projects are not that big, specially for the ones we have just created.
We do not need to worry about that.

hongkailiu commented:
@sjug 
In a training this week.
Will update early next week.

hongkailiu commented:
Done.
Moved and renamed to SetNamespaceLabels

With a little concern/question:
Update method uses the whole ns object as the posting body while patch method uses only the label/metadata part. Probably the latter is a bit more efficient.

I tried "oc label project ... --loglelvel=10" and saw only metadata is in the body of http-req.
That was why I used patch method in the first commit here.

Maybe namespaces/projects are not that big, specially for the ones we have just created.
We do not need to worry about that.

hongkailiu commented:
/test verify

sjug commented:
@hongkailiu 
I'm not sure how you define efficient...

In this solution `Get()` call returns a pointer not a new object. Then you're updating a field within the struct in that pointer, and calling an `Update()`. 

In your version you were creating new objects and then marshalling them to json, before making the `Patch()` call which will then need to unmarshall the json on the other end.

sjug commented on test/extended/cluster/cl.go:
```diff
@@ -79,6 +79,9 @@ var _ = g.Describe("[Feature:Performance][Serial][Slow] Load cluster", func() {
 				o.Expect(err).NotTo(o.HaveOccurred())
 				e2e.Logf("%d/%d : Created new namespace: %v", j+1, p.Number, nsName)
+				// label namespace nsName
+				_, err = SetNamespaceLabels(c, nsName, p.Labels)
```
If you check the other `Create{Configmaps,Secrets}` calls we like to check if the object exists before making the function call. I see that you have added the check in the `SetNamespaceLabels`, but could we (also?) have the check in line here (as below)?

hongkailiu commented:
Done.
Will clarify what I think about the patch and update offline.

sjug commented:
/lgtm

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

hongkailiu commented:
/test verify

sjug commented:
Hey @hongkailiu 

```
[WARNING] !!! gofmt needs to be run on the listed files
[FATAL] Try running 'gofmt -s -d [path]'
[FATAL] Or autocorrect with 'hack/verify-gofmt.sh | xargs -n 1 gofmt -s -w'
[ERROR] hack/verify-gofmt.sh exited with code 1 after 00h 00m 06s
```

hongkailiu commented:
/test cmd

hongkailiu commented:
/test extended_conformance_install

hongkailiu commented:
/test extended_conformance_install

sjug commented:
/LGTM

hongkailiu commented:
@smarterclayton @bparees
Let me know what I need to do before merging?
Thanks.

deads2k commented:
/approve

hongkailiu commented:
/test end_to_end

hongkailiu commented:
/test end_to_end

hongkailiu commented:
/test end_to_end

hongkailiu commented:
/test end_to_end
